tower-http/src/compression.rs

```diff
@@ -0,0 +1,214 @@
+use crate::common::*;
+use async_compression::tokio::bufread::BrotliEncoder;
+use async_compression::tokio::bufread::DeflateEncoder;
+use async_compression::tokio::bufread::GzipEncoder;
+use async_compression::tokio::bufread::ZstdEncoder;
+use futures_util::TryStreamExt;
+use http::HeaderMap;
+use hyper::Body;
+use std::io;
+use tokio_util::{
+    codec::{BytesCodec, FramedRead},
+    io::StreamReader,
+};
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct CompressionLayer {
+    _priv: (),
+}
+
+impl CompressionLayer {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl<S> Layer<S> for CompressionLayer {
+    type Service = Compression<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Compression { inner }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Compression<S> {
+    inner: S,
+}
+
+impl<ResBody, S> Service<Request<ResBody>> for Compression<S>
+where
+    S: Service<Request<ResBody>, Response = Response<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ResBody>) -> Self::Future {
+        let encoding = parse(req.headers()).unwrap_or(Encoding::Identity);
+
+        ResponseFuture {
+            future: self.inner.call(req),
+            encoding,
+        }
+    }
+}
+
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F> {
+    #[pin]
+    future: F,
+    encoding: Encoding,
+}
+
+impl<F, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<Body>, E>>,
```

Comment:
I think this is fine for now but I want to consider if we could maybe impl this over `impl Body<Buf = Bytes>` instead of `hyper::Body`. This way we could eventually add this into tonic quite easily. That is because in tonic we impl body like so https://docs.rs/tonic/0.4.0/tonic/body/struct.BoxBody.html#impl-Body